File: src/python/restxclient/restx_parameter.py
"""
RESTx: Sane, simple and effective data publishing and integration. 

Copyright (C) 2010   MuleSoft Inc.    http://www.mulesoft.com

This program is free software: you can redistribute it and/or modify 
it under the terms of the GNU General Public License as published by 
the Free Software Foundation, either version 3 of the License, or 
(at your option) any later version. 

This program is distributed in the hope that it will be useful, 
but WITHOUT ANY WARRANTY; without even the implied warranty of 
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
GNU General Public License for more details. 

You should have received a copy of the GNU General Public License 
along with this program.  If not, see <http://www.gnu.org/licenses/>. 

"""

"""
Definition of the L{RestxParameter} class.

This object describes the properties of a paremeter.
Parameters are used in different contexts.

    - A resource creation time parameter of a component.
    - A parameter for a component service
    - A parameter for a resource service (sub-resource)

"""

from restxclient.restx_client_exception import RestxClientException

def _numstr_to_num(x):
    """
    Translate a string to a numeric value.

    This sort of translation function is necessary to be more
    flexible in handling the type of input for parameter values.

    @param x:    String representing a number or a number
                 type.
    @type x:     string or number type

    @return:     Numeric value
    @rtype:      Integer or Float

    """
    if type(x) in [ int, float ]:
        return x
    try:
        return int(x)
    except:
        return float(x)

def _bool_convert(x):
    """
    Translate a string to a boolean value.

    This sort of translation function is necessary to be more
    flexible in handling the type of input for parameter values.

    Strings like 'y', 'yes', 'true', 't', '1' are interpreted
    as True, all other strings are considered to mean False.

    It works for upper and lower-case strings.

    @param x:     String representation of a boolean value
                  or a boolean.
    @type x:      string or boolean

    @return:      Boolean value.
    @rtype:       boolean

    """
    if type(x) is bool:
        return x
    if x.lower() in [ "y", "yes", "true", "t", "1" ]:
        return True
    else:
        return False


class RestxParameter(object):
    """
    Represents a parameter for a RESTx resource or service.

    This is a representation of the parameter's definition, not of the
    parameter's value.

    """
    # The keys to the parameter's meta data dictionary.
    __DESC_KEY     = "desc"
    __REQUIRED_KEY = "required"
    __TYPE_KEY     = "type"
    __DEFAULT_KEY  = "default"

    __name         = None
    __desc         = None
    __required     = None
    __default_val  = None
    __type_str     = None

    # All the known types identifiers for parameters
    __PARAM_STRING   = "string"
    __PARAM_PASSWORD = "password"
    __PARAM_BOOL     = "boolean"
    __PARAM_DATE     = "date"
    __PARAM_TIME     = "time"
    __PARAM_NUMBER   = "number"
    __PARAM_URI      = "uri"

    # This table can convert strings for specific types to a proper object
    # of the indicated type. The type identifier can be used to look up
    # a tuple, which contains suitable types as well as a suitable
    # conversion function. 'None' for the function indicates that no conversion
    # is necessary, since the underlying type is a string.
    __TYPE_CONVERT = {
        __PARAM_STRING   : ([str, unicode], None), 
        __PARAM_PASSWORD : ([str, unicode], None),
        __PARAM_BOOL     : ([str, unicode, bool], _bool_convert),
        __PARAM_DATE     : ([None], lambda x : date(*[ int(elem) for elem in x.split("-")])),
        __PARAM_TIME     : ([None], lambda x : time_class(*[ int(elem) for elem in x.split(":")])),
        __PARAM_NUMBER   : ([str, unicode, int, float], _numstr_to_num),
        __PARAM_URI      : ([str, unicode], None),
    }

    def __str__(self):
        """
        Return a string representation of this paramater.

        """
        return \
"""RestxParameter '%s':
    Description:   %s
    Type:          %s
    Required:      %s
    Default val:   %s""" % (self.__name, self.__desc, self.__type_str, self.__required, self.__default_val)

    def __init__(self, name, pdef):
        """
        Create a new parameter definition representation in memory.

        @param name:    Name of the parameter.
        @type name:     string

        @param pdef:    Dictionary with parameter definition. This is the
                        dictionary returned by the server when describing
                        a parameter.
        @type pdef:     dict

        """
        try:
            self.__name         = name
            self.__desc         = pdef[self.__DESC_KEY]
            self.__required     = _bool_convert(pdef.get(self.__REQUIRED_KEY, "n"))
            self.__type_str     = pdef[self.__TYPE_KEY].lower()

            if self.__type_str not in self.__TYPE_CONVERT:
                raise RestxClientException(("Server error: Type '%s' specified for parameter '%s', which is " +
                                          "not supported by this client library.") % (self.__type_str, self.__name))

            if self.__DEFAULT_KEY in pdef:
                # If a default value was specified, store it converted to
                # to the proper type.
                suitable_types, conversion_function = self.__TYPE_CONVERT[self.__type_str]
                value_str                           = pdef[self.__DEFAULT_KEY]
                if conversion_function:
                    self.__default_val = conversion_function(value_str)
                else:
                    self.__default_val = value_str
            else:
                if not self.__required:
                    raise RestxClientException("Server error: No default value specified for optional parameter '%s'." % self.__name)

        except KeyError, e:
            raise RestxClientException("Server error: Expected key '%s' missing in definition of parameter '%s'." % (str(e), self.__name))

    def sanity_check(self, value):
        """
        Check whether this is a valid value for this parameter.

        Raises an exception if there's a problem.

        @param value:      Some value object.
        @type value:       object

        """
        suitable_types, conversion_function = self.__TYPE_CONVERT[self.__type_str]
        if type(value) not in suitable_types:
            raise RestxClientException("Type '%s' is not suitable for parameter '%s'. Has to be one of '%s'." % (type(value), self.__name, suitable_types))

    def get_name(self):
        """
        Return the name of the parameter.

        @return:        Name of parameter.
        @rtype:         string

        """
        return self.__name

    def get_description(self):
        """
        Return the description of the parameter.

        @return:        Description of parameter.
        @rtype:         string

        """
        return self.__desc

    def get_parameter_type_str(self):
        """
        Return the string representation of the type of this parameter.

        @return:        String representation of parameter type.
        @rtype:         string

        """
        return self.__type_str

    def get_default_value(self):
        """
        Return the default value, if one was set.

        @return:        The default value, or None.
        @rtype:         Whichever type the default value has.

        """
        return self.__default_val

    def is_required(self):
        """
        Indicate whether this is a required parameter.

        @return:        Flag, which is True if required, False otherwise.
        @rtype:         boolean

        """
        return self.__required

    #
    # Some properties for conevenience
    #
    name        = property(get_name, None)
    description = property(get_description, None)
    required    = property(is_required, None)
    type_str    = property(get_parameter_type_str, None)

File: src/java/org/mulesoft/restx/clientapi/RestxParameter.java
/*      
 *  RESTx: Sane, simple and effective data publishing and integration. 
 *  
 *  Copyright (C) 2010   MuleSoft Inc.    http://www.mulesoft.com 
 *  
 *  This program is free software: you can redistribute it and/or modify 
 *  it under the terms of the GNU General Public License as published by 
 *  the Free Software Foundation, either version 3 of the License, or 
 *  (at your option) any later version. 
 * 
 *  This program is distributed in the hope that it will be useful, 
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of 
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
 *  GNU General Public License for more details. 
 * 
 *  You should have received a copy of the GNU General Public License 
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>. 
 */ 


package org.mulesoft.restx.clientapi;

import java.util.HashMap;
import java.lang.Number;
import java.math.BigDecimal;

public class RestxParameter
{
    protected final static String   DESC_KEY       = "desc";
    protected final static String   REQ_KEY        = "required";
    protected final static String   TYPE_KEY       = "type";
    protected final static String   DEFAULT_KEY    = "default";

    protected final static String   PARAM_STRING   = "string";
    protected final static String   PARAM_PASSWORD = "password";
    protected final static String   PARAM_BOOL     = "boolean";
    protected final static String   PARAM_DATE     = "date";
    protected final static String   PARAM_TIME     = "time";
    protected final static String   PARAM_NUMBER   = "number";
    protected final static String   PARAM_URI      = "uri";
    
    protected final static String[] KNOWN_TYPES    = { PARAM_STRING, PARAM_PASSWORD, PARAM_BOOL,
                                                       PARAM_NUMBER, PARAM_URI };
    protected final static String[] REQUIRED_KEYS  = { DESC_KEY, REQ_KEY, TYPE_KEY };
    
    protected String                name;
    protected String                desc;
    protected boolean               required;
    protected String                typeStr;
    protected Object                defaultVal;
    
    /**
     * Take an object (either boolean or string) and return boolean value.
     * 
     * If the value is a string and is either "y", "yes", "t", "true" or "1"
     * then this returns 'true', otherwise 'false'
     * 
     * @param  val  An object to be interpreted as a boolean value, either
     *              a boolean or a string representation of a boolean.
     * @return      A boolean value.
     * 
     * @throws RestxClientException 
     */
    protected boolean boolConvert(Object val) throws RestxClientException
    {
        if (val == null) {
            return false;
        }
        if (val.getClass() == Boolean.class) {
            return ((Boolean)val).booleanValue();
        }
        else if (val.getClass() == String.class) {   
            String valStr = ((String)val).toLowerCase();
            if (valStr.equals("y")  ||  valStr.equals("yes")  ||  valStr.equals("t")  ||
                valStr.equals("true")  ||  valStr.equals("1"))
            {
                return true;
            }
            return false;
        }
        else {
            throw new RestxClientException("Cannot convert boolean.");
        }
    }
    
    /**
     * Convert an object (either a String or a Number) to a Number.
     * 
     * If the value is a string then we return BigDecimal. Otherwise, it
     * returns the value object itself.
     * 
     * @param  val  Either a Number object or a String representing a number. 
     * @return      The corresponding Number object, either Integer, Float,
     *              Double or BigDecimal.
     * @throws      RestxClientException
     */
    protected Object numstrToNum(Object val) throws RestxClientException
    {
        if (val == null  ||  Number.class.isInstance(val)) {
            return val;
        }
        else if (val.getClass() == String.class) {
            return new BigDecimal((String)val);
        }
        else {
            throw new RestxClientException("Cannot convert number.");
        }
    }
        
    /**
     * Convert an object to the specified type.
     * 
     * The type is specified as a string (one of our known types).
     * The function then attempts to take the object, which may
     * be a string or other type and convert it to the specified
     * type.
     * 
     * @param  val       A value object.
     * @param  typeName  The name of the type. Has to be from
     *                   our KNOWN_TYPES list.
     * @return           Either String, Integer, Boolean, Float,
     *                   Double or BigDecimal.
     * @throws RestxClientException 
     */
    protected Object typeConvert(Object val, String typeName) throws RestxClientException
    {
        if (typeName.equals(PARAM_STRING)  ||  typeName.equals(PARAM_PASSWORD)  ||  typeName.equals(PARAM_URI)) {
            if (val.getClass() == String.class) {
                return val;
            }
        }
        else if (typeName.equals(PARAM_BOOL)) {
            return boolConvert(val);
        }
        else if (typeName.equals(PARAM_NUMBER)) {
            return numstrToNum(val);
        }
        throw new RestxClientException("Type value mismatch for value '" + val + "' and type '" + typeName + "'.");
    }
    

    /**
     * Construct a new RestxParameter.
     * 
     * This takes a parameter definition hash map as they would be received
     * straight from the server. No prior sanity checking needs to be performed
     * on it, since this constructor properly parses and interprets this map.
     * 
     * @param  name  Name of the new parameter.
     * @param  pdef  HashMap describing the parameter definition.
     * 
     * @throws       RestxClientException
     */
    public RestxParameter(String name, HashMap<String, ?> pdef) throws RestxClientException
    {
        this.name = name;
        try {
            // Sanity check on received information
            RestxServer.checkKeyset(pdef, REQUIRED_KEYS);

            desc     = (String)pdef.get(DESC_KEY);
            required = boolConvert(pdef.get(REQ_KEY));
            typeStr  = (String)pdef.get(TYPE_KEY);
            
            boolean validTypeName = false;
            for (String knownName : KNOWN_TYPES) {
                if (knownName.equals(typeStr)) {
                    validTypeName = true;
                    break;
                }
            }
            if (!validTypeName) {
                throw new RestxClientException("Type '" + typeStr + "' specified for parameter '" + name + 
                                             "', which is not supported by this client library.");
            }
            
            if (!required) {
                Object defaultValRepr = pdef.get(DEFAULT_KEY);
                if (defaultValRepr == null) {
                    throw new RestxClientException("No default value specified for optional parameter '" + name + ".");
                }
                defaultVal = typeConvert(pdef.get(DEFAULT_KEY), typeStr);
            }
            else {
                if (!required) {
                    throw new RestxClientException("No default value specified for optional parameter '" + name + ".");
                }
                defaultVal = null;
            }
        }
        catch (RestxClientException e) {
            throw new RestxClientException("Malformed parameter definition: " + e.getMessage());
        }
    }
    
    /**
     * Used to check whether a proposed value object is of the right type for this parameter.
     * 
     * Returns nothing if all is well, otherwise throws exception.
     * 
     * @param  val  Value object.
     * @throws      RestxClientException
     */
    public void sanityCheck(Object val) throws RestxClientException
    {
        typeConvert(val, typeStr);
    }

    /**
     * Returns the name of the parameter.
     * 
     * @return  Name of parameter.
     */
    public String getName()
    {
        return name;
    }
    
    /**
     * Returns the description of the parameter.
     * 
     * @return  Description of the parameter.
     */
    public String getDescription()
    {
        return desc;
    }
    
    /**
     * Returns the string that was used to define the type.
     * 
     * This is one of the KNOWN_TYPES.
     * 
     * @return Type string for this parameter.
     */
    public String getParameterTypeStr()
    {
        return typeStr;
    }
    
    /**
     * Returns the object holding the default value for this parameter.
     * 
     * Might be null if no default value was set.
     * 
     * @return The default value object for this parameter.
     */
    public Object getDefaultValue()
    {
        return defaultVal;
    }
    
    /**
     * Returns the flag indicating whether this is a required parameter.
     * 
     * @return The required flag.
     */
    public boolean isRequired()
    {
        return required;
    }
}
